concordium-std-derive/src/lib.rs

```diff
@@ -135,36 +157,61 @@ pub fn receive(attr: TokenStream, item: TokenStream) -> TokenStream {
 
     let ast: syn::ItemFn = syn::parse(item).expect("Receive can only be applied to functions.");
 
+    let mut setup_fn_optional_args = proc_macro2::TokenStream::new();
+    let mut fn_optional_args = vec![];
+
+    let amount_ident = format_ident!("amount");
+
+    if contains_attribute(&attrs, "payable") {
+        fn_optional_args.push(quote!(#amount_ident));
+    } else {
+        setup_fn_optional_args.extend(
+            quote!{
+                if #amount_ident.micro_gtu != 0 {
+                    return -1;
+                }
+            }
+        );
+    };
+
+    if contains_attribute(&attrs, "enable_logger") {
+        let logger_ident = format_ident!("logger");
+        setup_fn_optional_args.extend(
+            quote!(let mut #logger_ident = concordium_std::Logger::init();)
+        );
+        fn_optional_args.push(quote!(&mut #logger_ident));
+    }
```

Comment:
Isn't this very much the same as above for init. In which case, I think we should unify them in a function that returns two token streams.